frame/treasury/src/lib.rs

```diff
@@ -369,22 +374,55 @@ decl_module! {
 			Self::deposit_event(Event::<T>::Rejected(proposal_id, value));
 		}
 
-		/// Approve a proposal. At a later time, the proposal will be allocated to the beneficiary
-		/// and the original deposit will be returned.
+		/// Approve a proposal so that a user can call `payout_proposal` successfully.
 		///
 		/// # <weight>
 		/// - O(1).
 		/// - Limited storage reads.
-		/// - One DB change.
+		/// - Two DB changes.
 		/// # </weight>
 		#[weight = SimpleDispatchInfo::FixedOperational(100_000)]
 		fn approve_proposal(origin, #[compact] proposal_id: ProposalIndex) {
 			T::ApproveOrigin::try_origin(origin)
 				.map(|_| ())
 				.or_else(ensure_root)?;
 
-			ensure!(<Proposals<T>>::contains_key(proposal_id), Error::<T>::InvalidProposalIndex);
-			Approvals::mutate(|v| v.push(proposal_id));
+			let mut proposal = Proposals::<T>::get(proposal_id).ok_or(Error::<T>::InvalidProposalIndex)?;
+			let amount_approved = Self::amount_approved();
```

Comment:
would be better with a `mutate`.

Reply:
Ack, if you prefer the `try_mutate` syntax I can change it. But I want to confirm that logically/complexity it is the same thing as what I have written here.

Reply:
It's not that I "prefer" it. It is categorically worse.

Compare:
```
let foo = Foo::get();
foo = ...;
Foo::put(foo);
```

to:
```
Foo::mutate(|foo| {
    *foo = ...;
});
```

The second code has:
- fewer tokens (18 vs 15);
- far fewer identifiers (7 vs 4);
- no identifier duplication, whereas the first duplicates the `Foo` identifier.

These factors significantly impact readability, reviewability/auditability, maintainability and long-term, the likelihood of the code to be correct.

This is not abstract reasoning. I have seen bugs in the past within the pattern here. There are two bugs that are incipient to the first code and impossible in the second are a) mismatching identifiers for `put` and `get`, and b) forgetting the trailing `put`. I've definitely seen (b) in the wild.